### PR TITLE
Rework Persistent storage

### PR DESCRIPTION
**Pull request by petrspelos**
## Related Issue

By completely reworking the JSON data storage, I implemented a bunch of Error logging. This may help identify the issue in #75 .

## Changes
- Created an `ILogger` and `ConsoleLogger` to be able to log errors.
- Renamed `DataStorage` to `PersistentStorage`
- Completely reworked the implementation of JSONstorage.

The new implementation doesn't just store items into `collection/key`, it now uses the type as well.
When you store an instance of `MiunieUser` object and give it a group `TutorialServer` and key `user123456789` the directory structure will be:
```
Resources/MiunieUser/TutorialServer/user123456789.json
```
This allows us to retrieve all objects of a type. AND reuse the same group for other types.
Example:
```
Store(someUser, "TutorialServer", "u123456789");
```
Stores in `Resources/MiunieUser/TutorialServer/u123456789.json`
```
Store(guildConfiguration, "TutorialServer", "adminConfig"); // notice we use the same group.
```
Stores in `Resources/MiunieConfig/TutorialServer/adminConfig.json`

Previously you would have to use a different group for different types. Now it can be simplified.
```
var allUsersFromServer = RestoreMany<MiunieUser>("TutorialServer", "*");
var allConfigsFromServer = RestoreMany<MiunieConfig>("TutorialServer", "*");
```
Notice how the Group only represents the logical group, which in this case is the server name. That way we can seemingly get different type from the same group. (in reality those are separate subdirectories).

## Expected Feedback
This is quite a large refactoring, while I'm sure there is a number of small things that could be improved (and by all means point those out), I would appreciate if you could test this PR locally to make sure it work for you. :slightly_smiling_face: :tada:

### CODE REVIEW

File: src/Miunie.Core/Services/ProfileService.cs [hunk 1]
```diff
@@ -25,7 +28,11 @@ public async Task GiveReputation(MiunieUser invoker, MiunieUser target, MiunieCh
                 return;
             }
 
-            if (_reputationProvider.AddReputationHasTimeout(invoker, target)) { return; }
+            if (_reputationProvider.AddReputationHasTimeout(invoker, target))
+            {
+                _logger.Log($"User '{invoker.Name}' has a reputation timeout for User '{target.Name}', ignoring...");
```
Comment (Charly6596) [inline src/Miunie.Core/Services/ProfileService.cs]: This doesn't execute in my machine
image
The bot do nothing whenever i try to add rep to an user twice. With nothing i mean no output in discord, nor in the console

**Comment by svr333** [comment]
Did you execute it through Visual Studio?

File: src/Miunie.Logger/ConsoleLogger.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using Miunie.Core.Logging;
+
+namespace Miunie.Logger
+{
+    public class ConsoleLogger : ILogger
+    {
+        public void Log(string message)
+        {
+            if (message.StartsWith("ERROR"))
```
Comment (DraxCodes) [inline src/Miunie.Logger/ConsoleLogger.cs]: I feel like this should really be moved into a different method (LogError) for example. This sorta requires the user to already know that starting the log message with `ERROR` will format it differently.

File: src/Miunie.Storage/JsonPersistentStorage.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        }
+
+        private static string ToStoragePath(MemberInfo objType, string collection)
+            => Path.Combine(JsonDataDirectory, objType.Name, collection);
+
+        public void Store<T>(T obj, string collection, string key)
+        {
+            var path = ToStoragePath(typeof(T), collection);
+            var filePath = Path.Combine(path, $"{key}.json");
+            var json = JsonConvert.SerializeObject(obj);
```
Comment (DraxCodes) [inline src/Miunie.Storage/JsonPersistentStorage.cs]: Can we maybe add the pre-proccessor again for `IF DEBUG` we use `Formatting.Indentened`

File: src/Miunie.Logger/ConsoleLogger.cs (hunk 2 above)
Comment (petrspelos) [reply src/Miunie.Logger/ConsoleLogger.cs]: Absolutely :raised_hands: Thanks! :hearts:

File: src/Miunie.Storage/JsonPersistentStorage.cs (hunk 3 above)
Comment (petrspelos) [reply src/Miunie.Storage/JsonPersistentStorage.cs]: I'm on it, chief! :+1:

**Pull request merged**